File: src/main/java/gr/apt/controller/LeaveController.java
package gr.apt.controller;

import gr.apt.exception.LmsException;
import gr.apt.persistence.dto.ApproveLeaveDto;
import gr.apt.persistence.dto.LeaveDto;
import gr.apt.service.LeaveService;

import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import java.math.BigInteger;
import java.util.List;

@Path("/leave")
@Produces("application/json")
public class LeaveController {
    @Inject
    LeaveService service;

    @GET
    @Path("/list")
    public Response findAll(@QueryParam("index") Integer index, @QueryParam("size") Integer size) throws LmsException{
        List list = service.findAll(index,size);
        if(list != null){
            return Response.ok(list).build();
        }
        return Response.status(Response.Status.CONFLICT).build();
    }

    @GET
    @Path("/{personId}/list")
    public Response findAllByPersonId(@PathParam("personId") BigInteger personId,@QueryParam("index") Integer index, @QueryParam("size") Integer size) throws LmsException{
        List list = service.findAllByPersonId(personId,index,size);
        if(list != null){
            return Response.ok(list).build();
        }
        return Response.status(Response.Status.CONFLICT).build();
    }

    @GET
    @Path("/list-pending")
    public Response findAllPending(@QueryParam("index") Integer index, @QueryParam("size") Integer size) throws LmsException{
        List list = service.findAllPending(index,size);
        if(list != null){
            return Response.ok(list).build();
        }
        return Response.status(Response.Status.CONFLICT).build();
    }

    @GET
    @Path("/list-approved")
    public Response findAllApproved(@QueryParam("index") Integer index, @QueryParam("size") Integer size) throws LmsException{
        List list = service.findAllApproved(index,size);
        if(list != null){
            return Response.ok(list).build();
        }
        return Response.status(Response.Status.CONFLICT).build();
    }

    @GET
    @Path("/{id}")
    public Response findById(@PathParam("id")BigInteger id) throws LmsException{
        LeaveDto dto = service.findById(id);
        if(dto != null){
            return Response.ok(dto).build();
        }
        throw new LmsException("Could not find Leave with id:"+id);
    }

    @POST
    @Consumes("application/json")
    public Response create(LeaveDto dto) throws LmsException{
        return service.create(dto) ? Response.ok(dto).build():Response.status(Response.Status.CONFLICT).build();
    }

    @PUT
    @Consumes("application/json")
    public Response update(LeaveDto dto) throws LmsException{
        return service.update(dto) ? Response.ok(dto).build():Response.status(Response.Status.CONFLICT).build();
    }

    @DELETE
    @Consumes("application/json")
    public Response delete(LeaveDto dto) throws LmsException{
        return service.delete(dto) ? Response.ok().build():Response.status(Response.Status.CONFLICT).build();
    }

    @PUT
    @Path("/approve")
    @Consumes("application/json")
    public Response approve(ApproveLeaveDto dto) throws LmsException {
        return service.approve(dto) ? Response.ok(dto).build():Response.status(Response.Status.CONFLICT).build();
    }
}

File: src/main/java/gr/apt/lms/controller/LeaveController.kt
package gr.apt.lms.controller

import gr.apt.lms.dto.leave.ApproveLeaveDto
import gr.apt.lms.dto.leave.LeaveDto
import gr.apt.lms.exception.LmsException
import gr.apt.lms.service.LeaveService
import java.math.BigInteger
import javax.inject.Inject
import javax.ws.rs.*
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

@Path("/leave")
@Produces(MediaType.APPLICATION_JSON)
class LeaveController {
    @Inject
    lateinit var service: LeaveService

    @GET
    @Path("/list")
    @Throws(LmsException::class)
    fun findAll(@QueryParam("index") index: Int?, @QueryParam("size") size: Int?) =
        Response.ok(service.findAll(index, size)).build()


    @GET
    @Path("/{personId}/list")
    @Throws(LmsException::class)
    fun findAllByPersonId(
        @PathParam("personId") personId: BigInteger,
        @QueryParam("index") index: Int?,
        @QueryParam("size") size: Int?
    ) =
        Response.ok(service.findAllByPersonId(personId, index, size)).build()


    @GET
    @Path("/list-pending")
    @Throws(LmsException::class)
    fun findAllPending(@QueryParam("index") index: Int?, @QueryParam("size") size: Int?) =
        Response.ok(service.findAllPending(index, size)).build()


    @GET
    @Path("/list-approved")
    @Throws(LmsException::class)
    fun findAllApproved(@QueryParam("index") index: Int?, @QueryParam("size") size: Int?) =
        Response.ok(service.findAllApproved(index, size)).build()


    @GET
    @Path("/{id}")
    @Throws(LmsException::class)
    fun findById(@PathParam("id") id: BigInteger) = Response.ok(service.findById(id)).build()


    @POST
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun create(dto: LeaveDto) =
        Response.ok(service.create(dto)).build()


    @PUT
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun update(dto: LeaveDto) = Response.ok(service.update(dto)).build()

    @DELETE
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun delete(dto: LeaveDto) = Response.ok(service.delete(dto)).build()

    @PUT
    @Path("/approve")
    @Consumes(MediaType.APPLICATION_JSON)
    @Throws(LmsException::class)
    fun approve(dto: ApproveLeaveDto) = Response.ok(service.approve(dto)).build()
}